src/test/java/org/jenkinsci/plugins/workflow/graph/FlowNodeTest.java

```diff
@@ -0,0 +1,124 @@
+                    "  semaphore 'post-inner'\n" +
+                    "}\n" +
+                    "semaphore 'post-outer'\n" +
+                    "parallel a: {\n" +
+                    "  semaphore 'branch-a'\n" +
+                    "}, b: {\n" +
+                    "  semaphore 'branch-b'\n" +
+                    "}\n" +
+                    "semaphore 'last'", true));
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("pre-outer/1", b);
+                assertActiveSteps(b, "Start of Pipeline", "semaphore: pre-outer");
+                SemaphoreStep.success("pre-outer/1", null);
+                SemaphoreStep.waitForStart("pre-inner/1", b);
+            }
+        });
+        rr.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowRun b = rr.j.jenkins.getItemByFullName("p", WorkflowJob.class).getLastBuild();
+                assertActiveSteps(b, "Start of Pipeline", "stage: outer", "{ (outer)", "semaphore: pre-inner");
+                SemaphoreStep.success("pre-inner/1", null);
+                SemaphoreStep.waitForStart("inner/1", b);
+                assertActiveSteps(b, "Start of Pipeline", "stage: outer", "{ (outer)", "stage: inner", "{ (inner)", "semaphore: inner");
+                SemaphoreStep.success("inner/1", null);
+                SemaphoreStep.waitForStart("post-inner/1", b);
+                assertActiveSteps(b, "Start of Pipeline", "stage: outer", "{ (outer)", "semaphore: post-inner");
+                SemaphoreStep.success("post-inner/1", null);
+                SemaphoreStep.waitForStart("post-outer/1", b);
+                assertActiveSteps(b, "Start of Pipeline", "semaphore: post-outer");
+                SemaphoreStep.success("post-outer/1", null);
+                SemaphoreStep.waitForStart("branch-a/1", b);
+                SemaphoreStep.waitForStart("branch-b/1", b);
+            }
+        });
+        rr.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowRun b = rr.j.jenkins.getItemByFullName("p", WorkflowJob.class).getLastBuild();
+                // weird order caused by FlowGraphWalker DFS
+                assertActiveSteps(b, "{ (Branch: a)", "semaphore: branch-a", "Start of Pipeline", "parallel", "{ (Branch: b)", "semaphore: branch-b");
+                SemaphoreStep.success("branch-a/1", null);
+                SemaphoreStep.success("branch-b/1", null);
+                SemaphoreStep.waitForStart("last/1", b);
+                assertActiveSteps(b, "Start of Pipeline", "semaphore: last");
+                SemaphoreStep.success("last/1", null);
+                rr.j.waitForCompletion(b);
+                assertActiveSteps(b);
+            }
+        });
+    }
+    private static void assertActiveSteps(WorkflowRun b, String... expected) {
+        List<String> actual = new ArrayList<>();
```

Comment:
🐜  Did you mean to omit the second type?  I only ask because you made a point of adding it for the other ones.